pep-0612.rst

```diff
@@ -0,0 +1,258 @@
+
+Valid use locations
+^^^^^^^^^^^^^^^^^^^
+
+A declared ``ParameterSpecification`` can only be used in the place of the list of types in the declaration of a ``Callable`` type, or a user defined class which is generic in a ``ParameterSpecification`` variable (i.e., ``MyClass`` in the following example).
+
+.. code-block::
+
+   def foo(x: typing.Callable[TParams, int]) -> typing.Callable[TParams, str]: ... # Accepted
+   def foo(x: MyClass[TParams, int]) -> typing.Callable[TParams, str]: ... # Accepted
+   def foo(x: TParams) -> TParams: ... # Rejected
+   def foo(x: typing.List[TParams]) -> None: ... # Rejected
+   def foo(x: typing.Callable[[int, str], TParams]) -> None: ... # Rejected
+
+Semantics
+^^^^^^^^^
+
+The inference rules for the return type of a function invocation whose signature contains a ``ParameterSpecification`` variable are analogous to those around evaluating ones with ``TypeVar``\ s. 
+
+.. code-block::
+
+   def foo(x: typing.Callable[TParams, int]) -> typing.Callable[TParams, str]: ...
+   def bar(a: str, b: bool) -> int: ...
+   f = foo(bar) # f should be inferred to have the same signature as bar, but returning str
+   f("A", True) # Accepted
+   f(a = "A", b = True) # Accepted
+   f("A", "A") # Rejected
+
+Just as with traditional ``TypeVars``\ , a user may include the same ``ParameterSpecification`` multiple times in the arguments of the same function, to indicate a dependency between multiple arguments.  In these cases a type checker may choose to solve to a common behavioral supertype (i.e. a set of parameters for which all of the valid calls are valid in both of the subtypes), but is not obligated to do so.
+
+.. code-block::
+
+   def foo(
+       x: typing.Callable[TParams, int], y: typing.Callable[TParams, int]
+   ) -> typing.Callable[TParams, bool]: ...
+
+   def x_int_y_str(x: int, y: str) -> int: ...
+   def y_int_x_str(y: int, x: str) -> int: ...
+   foo(x_int_y_str, x_int_y_str) # Must return (x: int, y: str) -> int
+   foo(x_int_y_str, y_int_x_str) # Could return (__a: int, __b: str) -> int 
+                                 # This works because both callables have types that are behavioral
+                                 # subtypes of Callable[[int, str], int]
+
+   def keyword_only_x(*, x: int) -> int: ...
+   def keyword_only_y(*, y: int) -> int: ...
+   foo(keyword_only_x, keyword_only_y) # Must be rejected
+
+Use in ``Generic`` Classes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Just as with normal ``TypeVar``\ s, ``ParameterSpecification``\ s can be used to make generic classes as well as generic functions. These ought to be able to be mixed with normal ``TypeVar``\ s. This should also be made to work with protocols in the same manner.
+
+The components of a ``ParameterSpecification``
```

Comment:
I'm missing a brief description on what happens to keyword only ([PEP 3102](https://www.python.org/dev/peps/pep-3102/)) and to positional only ([PEP 570](https://www.python.org/dev/peps/pep-0570/)) arguments.

Will for example the following be accepted or rejected:
```
def add_logging(f: Callable[Ps, R]) -> Callable[Ps, R]:
       async def inner(*args: Ps.args, **kwargs: Ps.kwargs) -> R:
           # log
           return f(*args, **kwargs)
       return inner

@add_logging
def add(x:int, *, y: int = 0):
  return x + y

add(1, 2)
```
Note that today in Mypy, that `add` call (without the decorator) will be rejected.

I only found there's some talk abut it in the rejected proposals section.